README.md
# Abushakir (ባሕረ ሃሳብ) 
[![GitHub stars](https://img.shields.io/github/stars/Nabute/Abushakir?style=social)](https://github.com/Nabute/Abushakir/stargazers)       [![GitHub forks](https://img.shields.io/github/forks/Nabute/Abushakir?style=social)](https://github.com/Nabute/Abushakir/network)
[![GitHub license](https://img.shields.io/github/license/Nabute/Abushakir)](https://github.com/Nabute/Abushakir/blob/master/LICENSE)

The words Bahire Hasab originate from the ancient language of Ge'ez, ( Arabic: Abu Shakir) is a
time-tracking method, devised by the 12th pope of Alexandria, Pope St. Dimitri.

## What does it mean?

"Bahire Hasab /'bəhrɛ həsəb'/  " simply means "An age with a descriptive and chronological number". In some books it can also be found as "Hasabe Bahir", in a sense giving time an analogy, resembling a sea.

This package allows developers to implement Ethiopian Calendar and Datetime System in their application(s)`.

This package is implemented using the [UNIX EPOCH](https://en.wikipedia.org/wiki/Unix_time) which
means it's not a conversion of any other calendar system into Ethiopian, for instance, Gregorian Calendar.

Unix Epoch is measured using milliseconds since 01 Jan, 1970 UTC. In unix epoch leap seconds are ignored.

## Documentation
[Bahre Hasab](https://pub.dev/documentation/abushakir/latest/)

## Demo
[Ethiopian calendar](https://github.com/Nabute/ethiopian_calendar)

## Prerequisites

Before you begin, ensure you have met the following requirements:

* ```  equatable: ^2.0.3```


## Getting started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  bahere_hasab: ^0.0.2
```

In your library add the following import:

```dart
import 'package:bahere_hasab/bahere_hasab.dart';
```
## Example

```dart
   /**
   * Ethiopian Datetime Module [EtDatetime]
   */
  EtDatetime now = new EtDatetime.now(); // => 2012-07-28 17:18:31.466
  print(now.date); // => {year: 2012, month: 7, day: 28}
  print(now.time); // => {h: 17, m: 18, s: 31}

  EtDatetime covidFirstConfirmed = new EtDatetime(year: 2012, month: 7, day: 4);
  EtDatetime covidFirstConfirmedEpoch =
      new EtDatetime.fromMillisecondsSinceEpoch(covidFirstConfirmed.moment);

  EtDatetime covidFirstDeath = EtDatetime.parse("2012-07-26 23:00:00");

  /// Comparison of two EtDatetime Instances
  Duration daysWithOutDeath = covidFirstConfirmed.difference(covidFirstDeath);

  assert(daysWithOutDeath.inDays == -22, true); // 22 days

  assert(covidFirstDeath.isAfter(covidFirstConfirmed), true);

  assert(covidFirstDeath.isBefore(now), true);

  assert(covidFirstConfirmed.isAtSameMomentAs(covidFirstConfirmedEpoch), true);

  /**
   * Ethiopian Calendar Module [ETC]
   */
  ETC ethiopianCalendar = new ETC(year: 2012, month: 7, day: 4);

  ///
  print(ethiopianCalendar.monthDays(geezDay: true, weekDayName:true)); // Iterable Object of the given month
  print(ethiopianCalendar.monthDays().toList()[0]); // => [2012, 7, ፩, አርብ]
  // [year, month, dateNumber, dateNameIndex], Monday as First weekday

  print(ethiopianCalendar.nextMonth); // => ETC instance of nextMonth, same year
  print(ethiopianCalendar.prevYear); // => ETC instance of prevYear, same month

  /**
   * Bahire Hasab Module [BahireHasab]
   */
  BahireHasab bh = BahireHasab(year: 2011);
//  BahireHasab bh = BahireHasab(); // Get's the current year

  print(bh.getEvangelist(returnName: true)); // => ሉቃስ

  print(bh.getSingleBealOrTsom("ትንሳኤ")); // {month: ሚያዝያ, date: 20}

  bh.allAtswamat; // => List of All fasting and Movable holidays

  /**
   * Arabic or English number (1,2,3...) to Ethiopic or GE'EZ (፩, ፪, ፫...) number Converter
   */

  var input= [1, 10, 15, 20, 25, 78, 105, 333, 450, 600, 1000, 1001, 1010, 1056, 1200, 2013, 9999, 10000];

  for (var arabic in input) {
     print(ConvertToEthiopic(arabic)); // [፩, ፲, ፲፭, ፳, ፳፭, ፸፰, ፻፭, ፫፻፴፫, ፬፻፶, ፮፻, ፲፻, ፲፻፩, ፲፻፲, ፲፻፶፮, ፲፪፻, ፳፻፲፫, ፺፱፻፺፱, ፻፻]
  }

   /*
   * Conversion from any calendar (in this case, Gregorian) into Ethiopian Calendar.
   */
    DateTime gregorian1 = new DateTime.now();
    EtDatetime ethiopian1 = new EtDatetime.fromMillisecondsSinceEpoch(gregorian1.millisecondsSinceEpoch);

    print("Gregorian := ${gregorian1.toString()} is equivalent to Ethiopian ${ethiopian1.toString()}");
    // Gregorian := 2020-09-22 23:36:37.042962 is equivalent to Ethiopian 2013-01-12 20:36:37.042
    /*
    * Conversion from Ethiopian Calendar into any calendar (in this case, Gregorian).
    */
    EtDatetime ethiopian = new EtDatetime.now();
    DateTime gregorian = new DateTime.fromMillisecondsSinceEpoch(ethiopian.moment);

    print("Ethiopian ${ethiopian.toString()} is equivalent to Gregorian := ${gregorian.toString()}");
    // Ethiopian 2013-01-12 20:36:37.044 is equivalent to Gregorian := 2020-09-22 23:36:37.044

    print("Ethiopian EPOCH := ${ethiopian.moment}"); // Ethiopian EPOCH := 1600806997044
    print("Gregorian EPOCH := ${gregorian.millisecondsSinceEpoch}"); // Gregorian EPOCH := 1600806997044

```
For further implementation example see [this flutter application](https://github.com/Nabute/ethiopian_calendar)


## Contributors

Thanks to the following people who have contributed to this project:

* [@Nabute](https://github.com/Nabute) 📖
* [@Nahom](https://github.com/icnahom) 📖

<!---You might want to consider using something like the [All Contributors](https://github.com/all-contributors/all-contributors) specification and its [emoji key](https://allcontributors.org/docs/en/emoji-key).--->

## Contact

If you want to contact me you can reach me at <daniel@ibrave.dev>.

## License
<!--- If you're not sure which open license to use see https://choosealicense.com/--->

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE) file for details


CHANGELOG.md
## [0.0.2] - 2022-6-5

* Null-safety migration
## [0.0.1] - 2022-6-5




test/abushakir_test.dart
import 'package:test/test.dart';

import 'package:bahere_hasab/abushakir.dart';

void main() {
  group('Parameterized Constructors :', () {
    late EtDatetime ec;

    setUp(() {
      ec = EtDatetime(year: 2012, month: 07, day: 07);
    });

    test('Testing Year on Parameterized Constructor', () {
      expect(ec.year, 2012);
    });

    test('Testing Month on Parameterized Constructor', () {
      expect(ec.month, 07);
    });

    test('Testing Day on Parameterized Constructor', () {
      expect(ec.day, 07);
    });

    test('Testing Date on Parameterized Constructor', () {
      expect(ec.dayGeez, "፯");
    });
  });

  group('Parameterized Constructors (year only) :', () {
    late EtDatetime ec;

    setUp(() {
      ec = EtDatetime(year: 2010);
    });

    test('Testing Year on Parameterized Constructor', () {
      expect(ec.year, 2010);
    });

    test('Testing Month on Parameterized Constructor', () {
      expect(ec.month, 01);
    });

    test('Testing Day on Parameterized Constructor', () {
      expect(ec.day, 01);
    });
  });

  group('Named Constructors (.parse) :', () {
    late EtDatetime ec;

    setUp(() {
      ec = EtDatetime.parse("2012-07-07 15:12:17.500");
    });

    test('Should Print the parsed Date and Time', () {
      expect(ec.toString(), "2012-07-07 15:12:17.500");
    });

    test('Testing Year on Named Constructor', () {
      expect(ec.year, 2012);
    });

    test('Testing Month on Named Constructor', () {
      expect(ec.month, 07);
    });

    test('Testing Day on Named Constructor', () {
      expect(ec.day, 07);
    });

    test('Testing Date with Named Constructor', () {
      expect(ec.dayGeez, "፯");
    });

    test('Testing Hour on Named Constructor', () {
      expect(ec.hour, 15);
    });

    test('Testing Minute on Named Constructor', () {
      expect(ec.minute, 12);
    });

    test('Testing Second on Named Constructor', () {
      expect(ec.second, 17);
    });

    test('Testing Millisecond on Named Constructor', () {
      expect(ec.millisecond, 500);
    });
  });

  group('Named Constructors (.now) :', () {
    late EtDatetime ec;

    setUp(() {
      ec = EtDatetime.now();
    });
    test('Testing Minute on .now() Named Constructor', () {
      expect(ec.minute, DateTime.now().minute);
    });

    test('Testing Second on .now() Named Constructor', () {
      expect(ec.second, DateTime.now().second);
    });
  });

  group('Named Constructors (.fromMillisecondsSinceEpoch) :', () {
    late EtDatetime ec;

    // 1585731446021 == 2012-07-23 08:57:26.021
    setUp(() {
      ec = EtDatetime.fromMillisecondsSinceEpoch(1585731446021);
    });

    test('Should Print the parsed Date and Time', () {
      expect(ec.toString(), "2012-07-23 08:57:26.021");
    });

    test('Testing Year on Named Constructor', () {
      expect(ec.year, 2012);
    });

    test('Testing Month on Named Constructor', () {
      expect(ec.month, 07);
    });

    test('Testing Day on Named Constructor', () {
      expect(ec.day, 23);
    });

    test('Testing Date with Named Constructor', () {
      expect(ec.dayGeez, "፳፫");
    });

    test('Testing Hour on Named Constructor', () {
      expect(ec.hour, 08);
    });

    test('Testing Minute on Named Constructor', () {
      expect(ec.minute, 57);
    });

    test('Testing Second on Named Constructor', () {
      expect(ec.second, 26);
    });

    test('Testing Millisecond on Named Constructor', () {
      expect(ec.millisecond, 021);
    });
  });

  group('Named Constructors (.now) :', () {
    late EtDatetime ec;

    setUp(() {
      ec = EtDatetime.now();
    });
    test('Testing Minute on .now() Named Constructor', () {
      expect(ec.minute, DateTime.now().minute);
    });

    test('Testing Second on .now() Named Constructor', () {
      expect(ec.second, DateTime.now().second);
    });
  });

  group('Testing functions', () {
    late EtDatetime ec;
    setUp(() {
      // TODO: check if all constructors return same epoch.
//      ec = EtDatetime.now();
      ec = EtDatetime(year: 2012);
    });
    // 1585742246021 == 2012-07-23 11:57:26.021
    test('Testing isAfter()', () {
      print(ec.date);
      expect(ec.isAfter(EtDatetime(year: 2011)), true);
    });
    test('Testing isBefore()', () {
      expect(ec.isBefore(EtDatetime(year: 2080)), true);
    });
    test('Testing isAfter()', () {
      expect(
          ec.isAtSameMomentAs(EtDatetime(
              year: ec.year,
              month: ec.month,
              day: ec.day,
              hour: ec.hour,
              minute: ec.minute,
              second: ec.second,
              millisecond: ec.millisecond)),
          true);
    });
  });
  group('BahireHasab :', () {
    late BahireHasab bh;

    setUp(() {
      bh = BahireHasab(year: 2011);
    });

    test('Testing Abekte', () {
      expect(bh.abekte, 25);
    });

    test('Testing Metkih', () {
      expect(bh.metkih, 5);
    });

    test('Testing Nenewe', () {
      expect(bh.nenewe, {'month': 'የካቲት', 'date': 11});
    });

    // getSingleBealOrTsom
    test("Testing 'ነነዌ' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ነነዌ"), {'month': 'የካቲት', 'date': 11});
    });
    test("Testing 'ዓቢይ ጾም' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ዓቢይ ጾም"), {'month': 'የካቲት', 'date': 25});
    });

    test("Testing 'ደብረ ዘይት' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ደብረ ዘይት"), {'month': 'መጋቢት', 'date': 22});
    });

    test("Testing 'ሆሣዕና' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ሆሣዕና"), {'month': 'ሚያዝያ', 'date': 13});
    });

    test("Testing 'ስቅለት' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ስቅለት"), {'month': 'ሚያዝያ', 'date': 18});
    });

    test("Testing 'ትንሳኤ' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ትንሳኤ"), {'month': 'ሚያዝያ', 'date': 20});
    });

    test("Testing 'ርክበ ካህናት' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ርክበ ካህናት"), {'month': 'ግንቦት', 'date': 14});
    });

    test("Testing 'ዕርገት' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ዕርገት"), {'month': 'ግንቦት', 'date': 29});
    });

    test("Testing 'ጰራቅሊጦስ' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ጰራቅሊጦስ"), {'month': 'ሰኔ', 'date': 9});
    });

    test("Testing 'ጾመ ሐዋርያት' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ጾመ ሐዋርያት"), {'month': 'ሰኔ', 'date': 10});
    });

    test("Testing 'ጾመ ድህነት' on getSingleBealOrTsom", () {
      expect(bh.getSingleBealOrTsom("ጾመ ድህነት"), {'month': 'ሰኔ', 'date': 12});
    });
  });

  group('Number Convertor', () {
    late List<int> input;
    late List<String> output;
    setUp(() {
      input = [
        1,
        10,
        15,
        20,
        25,
        78,
        105,
        333,
        450,
        600,
        1000,
        1001,
        1010,
        1056,
        1200,
        2013,
        9999,
        10000
      ];
      output = [
        "፩",
        "፲",
        "፲፭",
        "፳",
        "፳፭",
        "፸፰",
        "፻፭",
        "፫፻፴፫",
        "፬፻፶",
        "፮፻",
        "፲፻",
        "፲፻፩",
        "፲፻፲",
        "፲፻፶፮",
        "፲፪፻",
        "፳፻፲፫",
        "፺፱፻፺፱",
        "፻፻"
      ];
    });
    test('Testing Convertor', () {
      for (var i = 0; i < input.length; i++) {
        expect(ConvertToEthiopic(input[i]), output[i]);
      }
    });
  });
}


lib/abushakir.dart
// Copyright (c) 2012 ETC or 2020 GC, Nabute and Nahom. All rights reserved.
// Use of this source code is governed by a MIT license that can be found
// in the LICENSE file.

///
/// Abushakir (Ge'ez: Bahire Hasab) is an arabic word that can be defined as a time-tracking method,
/// devised by the 12th pope of Alexandria, Pope St. Dimitri. St.
///
/// This library can be imported as
///
/// ```
/// import 'package:bahere_hasab/abushakir.dart';
/// ```
///
/// ## Ethiopian Date and Time
///
/// Use [EtDatetime] to represent moment in time.
/// You can also User [Duration] dart built it package to represent time span.
///
/// You can create EtDatetime object using one of the constructors or just by
/// parsing string with the correct format and also from UNIX timestamp.
///
/// ```
/// EtDatetime now = new EtDatetime.now();
/// EtDatetime covid19Confirmed = new EtDatetime(year: 2012, month: 7, day: 4);
///
/// EtDatetime emergencyLockdown = EtDatetime.parse("2012-07-26 23:00:00");
///
/// EtDatetime abushakirPublished = new EtDatetime.fromMillisecondsSinceEpoch(1586257594357)
/// ```
///
/// ## Ethiopian Calendar
///
/// Ethiopic year 1 E.E.3 starts on August 29, 8 C.E. (Julian), our R.D. 2796.
/// According to the definition of the "Ethiopic" calendar in Dershowitz and
/// Reingold's [Calendrical Calculations]
/// <https://www.amazon.com/Calendrical-Calculations-Nachum-Dershowitz/dp/0521702380>,
/// where it's the base calendar for
/// all computations.  See the book for algorithms for converting between
/// Ethiopic ordinals and many other calendar systems.
///
/// This calendar considers Monday as the beginning of the week. It does not do
/// any formatting, the base class, but any one who is working with this library
/// can use the output as the way they want to use.
///
/// You can create Ethiopian Calendar, ETC, object using one of the constructors
///
/// ```
/// ETC ethiopianCalendar = new ETC(year: 2012, month: 7, day: 4);
/// ETC today = ETC.today()
///
/// ```
///
///
/// ## Bahire Hasab
///
/// The words Bahire Hasab originate from the ancient language of Ge'ez,
/// ( Arabic: Abu Shakir) is a time-tracking method, devised by the 12th pope
/// of Alexandria, Pope St. Dimitri. This is the Ge'ez equivalent of Abushakir.
///
/// Bahire Hasab also helps man keep track of time, reminding him of the
/// Fasting seasons, prayer hours and Holidays.
///
/// You can create BahireHasab object using constructor
///
/// ```
/// BahireHasab someRandomYear = BahireHasab(year: 2011);
///
/// // Year parameter is optional. If year is not provided, it will assume the
/// current year.
///
/// BahireHasab thisYear = BahireHasab();
///
/// ```
///
library abushakir;

import 'dart:convert';
import 'package:equatable/equatable.dart';

part 'src/etDateTime.dart';

part 'src/calendar/bahireHasab.dart';

part 'src/utils/constants.dart';

part 'src/utils/calendar_exceptions.dart';

part 'src/utils/convertor.dart';

part 'src/calendar/calendar.dart';

part 'src/calendar/abushakir.dart';


example/example.dart
// Copyright 2020 GC (2012 ETC) Nabute and Nahom. All rights reserved.
// Use of this source code is governed by MIT license, which can be found
// in the LICENSE file.

import 'package:abushakir/bahere_hasab.dart';

/// An Example of using the package to create and manipulate Ethiopian Date and
/// Time with the unique Calendar system which includes the way ethiopians
/// use to find movable feasts and holiday.

// import 'package:bahere_hasab/abushakir.dart';

void main() {
  /**
   * Ethiopian Datetime Module [EtDatetime]
   */
  EtDatetime now = new EtDatetime.now(); // => 2012-07-28 17:18:31.466
  print(now.date); // => {year: 2012, month: 7, day: 28}
  print(now.time); // => {h: 17, m: 18, s: 31}

  EtDatetime covidFirstConfirmed = new EtDatetime(year: 2012, month: 7, day: 4);
  EtDatetime covidFirstConfirmedEpoch =
      new EtDatetime.fromMillisecondsSinceEpoch(covidFirstConfirmed.moment);

  EtDatetime covidFirstDeath = EtDatetime.parse("2012-07-26 23:00:00");

  /// Comparison of two EtDatetime Instances
  Duration daysWithOutDeath = covidFirstConfirmed.difference(covidFirstDeath);

  assert(daysWithOutDeath.inDays == -22, true); // 22 days

  assert(covidFirstDeath.isAfter(covidFirstConfirmed), true);

  assert(covidFirstDeath.isBefore(now), true);

  assert(covidFirstConfirmed.isAtSameMomentAs(covidFirstConfirmedEpoch), true);

  /**
   * Ethiopian Calendar Module [ETC]
   */
  ETC ethiopianCalendar = new ETC(year: 2011, month: 13, day: 4);

  ///
  print(ethiopianCalendar.monthDays(
      geezDay: true, weekDayName: true)); // Iterable Object of the given month
  print(ethiopianCalendar.monthDays().toList()[0]); // => [2012, 7, 1, 1]
  // [year, month, dateNumber, dateNameIndex], Monday as First weekday

  print(ethiopianCalendar.nextMonth); // => ETC instance of nextMonth, same year
  print(ethiopianCalendar.prevYear); // => ETC instance of prevYear, same month

  /**
   * Bahire Hasab Module [BahireHasab]
   */
  BahireHasab bh = BahireHasab(year: 2011);
//  BahireHasab bh = BahireHasab(); // Get's the current year

  print(bh.getEvangelist(returnName: true)); // => ሉቃስ

  print(bh.getSingleBealOrTsom("ትንሳኤ")); // {month: ሚያዝያ, date: 20}

  bh.allAtswamat; // => List of All fasting and Movable holidays

  /**
   * Arabic or English number (1,2,3...) to Ethiopic or GE'EZ (፩, ፪, ፫...) number Converter
   */

  var input = [
    1,
    10,
    15,
    20,
    25,
    78,
    105,
    333,
    450,
    600,
    1000,
    1001,
    1010,
    1056,
    1200,
    2013,
    9999,
    10000
  ];

  for (var arabic in input) {
    print(ConvertToEthiopic(
        arabic)); // [፩, ፲, ፲፭, ፳, ፳፭, ፸፰, ፻፭, ፫፻፴፫, ፬፻፶, ፮፻, ፲፻, ፲፻፩, ፲፻፲, ፲፻፶፮, ፲፪፻, ፳፻፲፫, ፺፱፻፺፱, ፻፻]
  }

  /*
  Conversion from any calendar (in this case, Gregorian) into Ethiopian Calendar.
   */
  DateTime gregorian1 = new DateTime.now();
  EtDatetime ethiopian1 = new EtDatetime.fromMillisecondsSinceEpoch(
      gregorian1.millisecondsSinceEpoch);
  print(
      "*******************   Gregorian into Ethiopian   ***********************");
  print(
      "Gregorian := ${gregorian1.toString()} is equivalent to Ethiopian ${ethiopian1.toString()}");

  /*
   Conversion from Ethiopian Calendar into any calendar (in this case, Gregorian).
   */
  EtDatetime ethiopian = new EtDatetime.now();
  DateTime gregorian =
      new DateTime.fromMillisecondsSinceEpoch(ethiopian.moment);
  print("Ethiopian EPOCH := ${ethiopian.moment}");
  print("Gregorian EPOCH := ${gregorian.millisecondsSinceEpoch}");
  print(
      "*******************   Ethiopian into Gregorian   ***********************");
  print(
      "Ethiopian ${ethiopian.toString()} is equivalent to Gregorian := ${gregorian.toString()}");
}
